### PR TITLE
modfied blender data parser

### PR DESCRIPTION
**Pull request by hardikdava**
@yzslab This PR enables following features.

- enables nerfstudio data through NGP data parser
- added support for auto image size detection for blender data parser
- added support for different focal length for blender data parser
- readme updated

### CODE REVIEW

File: internal/dataparsers/blender_dataparser.py [hunk 1]
```diff
@@ -65,17 +63,26 @@ def _parse_transforms_json(self, split: str) -> ImageSet:
...
+
         # parse focal length
         fx = torch.tensor(
             [fov2focal(fov=transforms["camera_angle_x"], pixels=width)],
             dtype=torch.float32,
         ).expand(R.shape[0])
-        # TODO: allow different fy
-        fy = torch.clone(fx)
+        fy = torch.tensor(
+            [fov2focal(fov=transforms["camera_angle_y"], pixels=height)],
```
Comment (yzslab) [inline internal/dataparsers/blender_dataparser.py:80]: Have you tested it? Looks like the nerf synthetic dataset does not have `camera_angle_y`.

File: internal/dataparsers/ngp_dataparser.py [hunk 2]
```diff
@@ -152,14 +152,17 @@ def generate_image_set(indices: list):
...
 
         if self.config.pcd_from == "file" or self.config.pcd_from == "auto" and is_pcd_file_exist is True:
             print(f"Load pcd from '{pcd_file_path}'")
-            from internal.utils.graphics_utils import fetch_ply
-            pcd = fetch_ply(pcd_file_path)
+            from internal.utils.graphics_utils import fetch_ply_without_rgb_normalization
+            pcd = fetch_ply_without_rgb_normalization(pcd_file_path)
             point_cloud = PointCloud(
                 xyz=pcd.points,
                 rgb=(pcd.colors * 255).astype(np.uint8),
```
Comment (yzslab) [inline internal/dataparsers/ngp_dataparser.py:168]: You should update this line too.

**Review by yzslab — CHANGES_REQUESTED** [review]

File: internal/dataparsers/ngp_dataparser.py (hunk 2 above)
Comment (yzslab) [reply internal/dataparsers/ngp_dataparser.py:168]: The `pcd.colors` should not be multiplied by 255 now.

File: internal/dataparsers/blender_dataparser.py (hunk 1 above)
Comment (hardikdava) [reply internal/dataparsers/blender_dataparser.py:80]: @yzslab just checked and you are right. We have to use `camera_angle_x` only. So in theory it has same height and width.

File: internal/dataparsers/blender_dataparser.py [hunk 3]
```diff
@@ -65,17 +63,26 @@ def _parse_transforms_json(self, split: str) -> ImageSet:
...
         T = world_to_camera[:, :3, 3]
 
+        height, width = None, None
+        for image_path in image_path_list:
+            img = cv2.imread(image_path)
+            if img is None:
+                continue
+            height, width = img.shape[:2]
+            break
+
```
[the remark this answers is not in the document]
Comment (hardikdava) [reply internal/dataparsers/blender_dataparser.py]: fixed.

File: internal/dataparsers/blender_dataparser.py (hunk 1 above)
Comment (hardikdava) [reply internal/dataparsers/blender_dataparser.py]: fixed.

File: internal/dataparsers/ngp_dataparser.py (hunk 2 above)
Comment (hardikdava) [reply internal/dataparsers/ngp_dataparser.py]: fixed.

**Comment by yzslab** [comment]
Does the NGP dataparser work as you expected? One reason that I did not introduce it in the README because I have not really tested it. Another problem is that most of the NGP datasets contain camera distortions. Theoretically this can be solved with the undistortion functions provided by OpenCV. But comparing with the colmap undistorter, OpenCV does not perform well on the experiment metrics. So I hid it in case someone use it without being aware of the distortion problem.

**Comment by hardikdava** [comment]
@yzslab yes, you are right. But nerfstudio uses `cv2.undistort` function to undistort images and it is working very well.

**Comment by yzslab** [comment]
> @yzslab yes, you are right. But nerfstudio uses `cv2.undistort` function to undistort images and it is working very well.

When did you test it, and how well it worked?

It has been a confirmed issue since January: https://github.com/nerfstudio-project/nerfstudio/issues/2849, and was only claimed to have been fixed two weeks ago: https://github.com/nerfstudio-project/nerfstudio/pull/3382.
But not sure whether it was fixed really because they did not provide any metrics.
Even the team members of NeRFStudio can not clearly tell what makes such a difference.

**Comment by yzslab** [comment]
If you really need it, and know how to deal with the distortion, it still can be enabled with the full class path, e.g., `--data.parser internal.dataparsers.ngp_dataparser.NGP`, without being added to `__init__.py`.

Besides, there is a better way to dealing with the NGP dataset. You can convert it back to colmap sparse model like [utils/meganerf2colmap.py](https://github.com/yzslab/gaussian-splatting-lightning/blob/main/utils/meganerf2colmap.py). Not only can you utilize colmap's undistorter but also a better point cloud that some of the NGP datasets do not contain.

**Comment by hardikdava** [comment]
@yzslab Please review this PR.

**Comment by yzslab** [comment]
Thanks!

**Pull request merged**